src/init/mod.rs

```diff
@@ -17,12 +17,59 @@ using whatever mechanism is available in the host shell--this two-phase solution
 has been developed as a compatibility measure with `eval $(starship init X)`
 */
 
-fn path_to_starship() -> io::Result<String> {
-    let current_exe = env::current_exe()?
-        .to_str()
-        .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "can't convert to str"))?
-        .to_string();
-    Ok(current_exe)
+struct StarshipPath {
+    native_path: PathBuf,
+}
+impl StarshipPath {
+    fn init() -> io::Result<Self> {
+        Ok(Self {
+            native_path: env::current_exe()?,
+        })
+    }
+    fn str_path(&self) -> io::Result<&str> {
+        let current_exe = self
+            .native_path
+            .to_str()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "can't convert to str"))?;
+        Ok(current_exe)
+    }
+    fn sprint(&self) -> io::Result<String> {
+        self.str_path().map(|s| s.replace("\"", "\"'\"'\""))
+    }
+    fn sprint_posix(&self) -> io::Result<String> {
+        // On non-Windows platform, return directly.
+        if cfg!(not(target_os = "windows")) {
+            return self.sprint();
+        }
+        let str_path = self.str_path()?;
+        let res = std::process::Command::new("cygpath.exe")
+            .arg(str_path)
+            .output();
+        let output = match res {
+            Ok(output) => output,
+            // Failed to execute cygpath.exe means there're not inside cygwin evironment,return directly.
+            Err(_) => {
+                return self.sprint();
+            }
```

Comment:
Might also make sense to combine with the match below.

```suggestion
            Err(e) if e.kind() == ErrorKind::NotFound => {
                return self.sprint();
            }
```